### ISSUE TITLE
Using BalancedConsumer with consumer_timeout_ms options do not receive message

### ISSUE THREAD

**Issue by gossel-j** [Parsely/pykafka#528]
When I use BalancedConsumer with option consumer_timeout_ms, when I try to connect another BalancedConsumer it stops receive message and then timeout.

So basically after a rebalancing, the consumer never give new message and then timeout, only appear with the option consumer_timeout_ms.

```python
#!/usr/bin/env python

from argparse import ArgumentParser
import json
import logging
from pykafka import KafkaClient


def consume(args):
    client = KafkaClient(
        hosts=args.broker,
        zookeeper_hosts=args.zookeeper,
    )
    topic = client.topics[args.topic]
    consumer = topic.get_balanced_consumer(
        args.group,
        managed=False,
        auto_commit_enable=True,
        auto_start=True,
        consumer_timeout_ms=60000, # if this line is commented it works great
    )

    for msg in consumer:
        if msg is None:
            break
        logging.info("partiton=%s offset=%s value=%s" % (msg.partition.id, msg.offset, msg.value))


def main():
    logging.basicConfig(level=logging.INFO)
    parser = ArgumentParser(description='Produce dummy messages to kafka')
    parser.add_argument(
        '--broker', '-b',
        dest='broker',
        default='127.0.0.1:9092',
        type=str,
        metavar='kafka_broker_address(es)',
    )
    parser.add_argument(
        '--zookeeper', '-z',
        dest='zookeeper',
        default='127.0.0.1:2181',
        type=str,
        metavar='zookeeper_address(es)',
    )
    parser.add_argument(
        '--topic', '-t',
        dest='topic',
        default='dummy',
        type=str,
        metavar='topic_name',
    )
    parser.add_argument(
        '--group', '-g',
        dest='group',
        type=str,
        metavar='group_name',
        required=True,
    )

    args = parser.parse_args()

    consume(args)


if __name__ == '__main__':
    main()
```

**PyKafka version**: 2.3.1
**Kafka version**: 8.2.2

**Comment by emmett9001** [comment]
Thanks @gossel-j. At first glance it looks possible that this is a real issue. I'll look into it more deeply.

**Comment by emmett9001** [comment]
I haven't gone super deep on this, but it looks like this might he happening because [this line](https://github.com/Parsely/pykafka/blob/40f41021adb25f2ab21455c1ecb869c96b9f608c/pykafka/simpleconsumer.py#L414) is blocking for longer than the rebalance takes to complete. @gossel-j do you have any log output indicating what your timeout looks like?

**Comment by emmett9001** [comment]
I'm going deep on this today. What I've found so far is this sequence of events:

* `BalancedConsumer` running normally
* In normal operation, a `BalancedConsumer.consume()` call's internal call to `SimpleConsumer.consume()` waits [here](https://github.com/Parsely/pykafka/blob/6e4698b06571e83f5cecb4a5c94b7514df2c8137/pykafka/simpleconsumer.py#L421) for a new message to arrive. It waits for `_consumer_timeout_ms`.
* While the `SimpleConsumer.consume()` call is blocking, a rebalance is triggered.
* The `consume()` call waits for `_consumer_timeout_ms` and continues. Since a rebalance is in progress, [this line](https://github.com/Parsely/pykafka/blob/6e4698b06571e83f5cecb4a5c94b7514df2c8137/pykafka/simpleconsumer.py#L433) or the matching line above is reached, causing `ConsumerStoppedException`.
* This causes the `BalancedConsumer` to [immediately re-loop](https://github.com/Parsely/pykafka/blob/6e4698b06571e83f5cecb4a5c94b7514df2c8137/pykafka/balancedconsumer.py#L732) and [recheck the timeout condition](https://github.com/Parsely/pykafka/blob/6e4698b06571e83f5cecb4a5c94b7514df2c8137/pykafka/balancedconsumer.py#L725). This condition is `True` since the `consume()` call waited for `_consumer_timeout_ms`.
* At this point, the `BalancedConsumer.consume()` call returns `None`, which is not the correct behavior.

My thought on how to fix this behavior is that there is one basic thing going wrong here: the semaphore condition in `SimpleConsumer.consume()` should not continue to wait when the consumer itself has been stopped. Doing so makes no sense, since `stop()` stops new messages from being fetched. The fix, then, should be to release the semaphore on `stop()` to unblock any so blocked calls to `consume()`.

Applying this fix reveals another issue. Once the `BalancedConsumer` is able to wait on its internal consumer for less than `_consumer_timeout_ms`, the re-loop may still be calling `consume()` on the old `self._consumer`. The rebalancing logic may not have reassigned `self._consumer`. It has [a lot to do](https://github.com/Parsely/pykafka/blob/6e4698b06571e83f5cecb4a5c94b7514df2c8137/pykafka/balancedconsumer.py#L359-L376) between `stop()`ing the consumer and reassigning `self._consumer`, so this is a common occurrence. Since `self._consumer` [here](https://github.com/Parsely/pykafka/blob/6e4698b06571e83f5cecb4a5c94b7514df2c8137/pykafka/balancedconsumer.py#L728) is still pointing to the old, stale `SimpleConsumer`, and since the semaphore was only released once by `stop()`, [this logic](https://github.com/Parsely/pykafka/blob/6e4698b06571e83f5cecb4a5c94b7514df2c8137/pykafka/simpleconsumer.py#L421) blocks once again.

The fix here is to ensure that `self._consumer.consume` cannot be called during a rebalance, so we're always sure that the `self._consumer` we're referencing is the currently valid one.

**Comment by emmett9001** [comment]
@gossel-j Please comment here if the fix in https://github.com/Parsely/pykafka/pull/576 doesn't work for you.

**Comment by gossel-j** [comment]
@emmett9001 Looks like working great when I use pykafka from master now. Thanks for fixing this !
